### ISSUE TITLE
Get error using useMediaQuery in SSR

### ISSUE THREAD

**Issue by unix** [zeit-ui/react#258]
# Bug report 🐞

When you use `useMediaQuery` in `next.js` or other SSR frameworks, you may get an error:

```js
// or other render errors
Prop `style` did not match!
```

Actually, it's not a bug, this issue will explain how to handle them in SSR.

### Related issues

  - [Problem with useMediaQuery hook](https://spectrum.chat/zeit-ui/general/problem-with-usemediaquery-hook~da638e09-9336-41b9-97ce-db4b5877afe2)

**Comment by unix** [comment]
## Error source

The error comes from the SSR frameworks, because the current DOM is not the same as that rendered by the SSR frameworks. The problem usually happens before rendering, we manually modify the dom.

For example, when you jugde whether to display components based on the value of `localStorage`(Browser API), the resulting DOM structure will be completely different from the server-side rendering. Because the server does not have this API, it cannot render the same dom:

```jsx
() => {
  const [state, setState] = useState(flase)
  if (typeof window !== 'undefined') {
    setState(window.localStorage.getItem('showWorld'))
  }
  return (
    <p>Hello {state ? <span>World</span> : null}</p>
  )
}
```

<br />

## For `zeit-ui/react`

In `zeit-ui`, some functions only work on the browser side, `useMediaQuery` is a typical exmaple.
This is not a technical reason or our components are not good enough, because media queries are browser specific, only when we are in the browser environment can we know the user's device information.

Strictly speaking, this is based on the needs of the client and can only be rendered by the client.

<br/>

## Solutions

Existing solutions can be divided into two categories:

  - **Client re-render**: Ignore all client information when rendering on the server, render 2 times after client obtains HTML string. Because of the need for re-render, this solution will be slower, but the speed of render is in an order of magnitude.
  - **User agent**: The server obtains the device information of each user and customizes different DOM structures, its advantage is faster. But this means that all HTML strings (first bytes) are dynamic, your basic HTML **cannot be cached**.

### Client re-render

You can execute the client's code after first rendering:

```jsx
const isXS = useMediaQuery("xs")

useEffect(() => {
  setMyState1(isXS)
  setMyState2(window.localStorage.getItem('showWorld'))
}, [])
```

This code may cause display confusion or screen flicker (when you change DOMs), because it triggers 2 renderings in a very short time, this problem can be solved by hiding the first rendered page:

```jsx
const [showPage, setShowPage] = setState(false)
const isXS = useMediaQuery("xs")

useEffect(() => {
  setMyState1(isXS)
  setShowPage(true)
}, [])

if (!showPage) return null
return <div>...</div>
```

Unfortunately, such code will make the server render fixed HTML strings. Because the SSR framework does not work properly, all the first bytes will be the same, this will cause your page to lose SEO function.

The better solution is to reference `children` in a fixed way, but do not show them before client rendering: (The first HTML string is prepared for spiders. It is not displayed by default)

```jsx
const [showPage, setShowPage] = setState(false)
const isXS = useMediaQuery("xs")

useEffect(() => {
  setMyState1(isXS)
  setShowPage(true)
}, [])

if (!showPage) return <div style={{ display: 'none' }}>{children}</div>
return <div>...</div>
```

**TLDR**: `zeit-ui/react` provide a simpler solution, the `Page` component can block the first rendering by default, just set the `render` prop.

A complete example for `useMediaQuery` and `Page`: (view on [codesandbox](https://codesandbox.io/s/zeit-ui-react-client-render-ef45k?file=/pages/index.js))

```jsx
const Home = () => {
  const isXS = useMediaQuery('xs')
  const [showSpan, setShowSpan] = useState(false)
  useEffect(() => {
    setShowSpan(isXS)
  }, [isXS])

  return (
    <Page render="effect-seo">
      {showSpan && <span>is xs</span>}
    </Page>
  )
}
```

Or you don't care about SEO, it's easier:

```jsx
const Home = () => {
  const isXS = useMediaQuery('xs')

  return (
    <Page render="effect">
      {isXS && <span>is xs</span>}
    </Page>
  )
}
```

### User agent

This is a best practice, **but the first HTML of the page will not be cached**. We can return different HTML strings according to different devices of users:

```jsx
import parser from 'ua-parser-js'
import mediaQuery from 'css-mediaquery'

const deviceType = parser(req.headers['user-agent']).device.type || 'desktop'
  const ssrMatchMedia = query => ({
    matches: mediaQuery.match(query, {
      width: deviceType === 'mobile' ? '0px' : '1024px',
    }),
  })

const Home = () => {
  cosnt something = useMediaQuery('xs', { ssrMatchMedia })
  return <div>...</div>
}
```

This is a simple example, In fact, you have to do more than that. Its principle is to obtain the device information through the `header`, and then return the specified HTML string.

**Note:** The content of the page depends on the calculation of the server, If you are using serverless, **each page will be treated as a serverless**. (The Vercel limits each project to a maximum of 12 serverless, Other platforms have similar limits)

<br />

### Balance

1. If you want to solve it in simplest way, try `Page` component and `React.useEffect`.
2. If you need a little customization, add several `React.useEffect` is a good choice.
3. If you need a higher first screen speed, use `User agent` plan. (Because it can't be cached, it can't be ensured that the speed of this scheme must be higher.)

**Comment by liesislukas** [comment]
@unix your English is good enough, just continue writing useful content 💯
